### ISSUE TITLE
Add custom icons to unity atoms assets

### ISSUE THREAD

**Issue by AdamRamberg** [AdamRamberg/unity-atoms#23]
Wouldn't this library be soooo much better with gizmos? :D 

Investigate if it is possible to just use one image / icon and use it for all script icons in the inspector and all unity atom assets created (like the Post Processing Stack is doing - https://github.com/Unity-Technologies/PostProcessing)

**Comment by jeffcampbellmakesgames** [comment]
I think I hear what you're saying, although I think there might be a bit of confusion between [Gizmos](https://docs.unity3d.com/ScriptReference/Gizmos.html) which are intended for visual debugging and setup in the Scene View versus overriding the default `ScriptableObject` icon so that unity-atom assets are visually distinguished. 

This is pretty easy to setup. Given the number of distinct `ScriptableObject` derived types it would probably help to create a table/spreadsheet of each of the types we want to assign an icon to. That way we can infer a semantic or general intent thats easily visualized for each one and then create/assign an icon for it.

**Comment by AdamRamberg** [comment]
Got you! Was a little bit lazy and didn't do enough research before making the issue (just wanted to create it to keep track of it). Changed the name of the issue sto make it less confusing. 

A spreadsheet / table sounds like a good idea 👌My first initial thought is to create an icon per type, Variable, Event, Listeners, MonoHooks, Functions and Actions. How does that sound to you? Wouldn't mind if you start on a table. 

Another suggestion would be to use the icon that I used for my blog post [here](https://medium.com/@adamramberg/unity-atoms-tiny-modular-pieces-utilizing-the-power-of-scriptable-objects-e8add1b95201) (first picture). Then we could just change the color for each type. But again, would like some input on this.

**Comment by AdamRamberg** [comment]
@jeffcampbellmakesgames How do you add custom icons to scripts / assets? From looking at your [repo](https://github.com/jeffcampbellmakesgames/unity-package-tools) it seems like meta data is pointing to the guid of the icon. However, do one need to manually set this guid in the meta file(s) manually or is there some other way?

**Comment by jeffcampbellmakesgames** [comment]
You can either add that yaml section manually through the meta file or finding the script in the Unity Editor and manually assigning it through the inspector.

In the screenshot below, selecting the script's icon in the inspector and then selecting `Other` will allow you to manually select the icon to use.

![image](https://user-images.githubusercontent.com/1663648/59336575-1ba72a00-8cff-11e9-86d9-e687238088ab.png)

**Comment by soraphis** [comment]
You can create a folder "Gizmos" and in there an image: `<NameOfYourClass> Icon.png`
(note the space between the class name and "Icon").

Assets of this class type should automatically use this icon

[source](https://forum.unity.com/threads/custom-scriptableobject-icons-thumbnail.256246/#post-1693711)

**Comment by jeffcampbellmakesgames** [comment]
> You can create a folder "Gizmos" and in there an image: `<NameOfYourClass> Icon.png`
> (note the space between the class name and "Icon").
> 
> Assets of this class type should automatically use this icon
> 
> [source](https://forum.unity.com/threads/custom-scriptableobject-icons-thumbnail.256246/#post-1693711)
> 
> edit: reading further in this thread, it seems in newer versions you need to take the namespace into account, see this post: https://forum.unity.com/threads/custom-scriptableobject-icons-thumbnail.256246/#post-3654346

Just to help clarify, Gizmos are for in-sceneview icons while the types of icons these would be for would be on the ScriptableObjects in the project view or wherever they are assigned as a serialized reference.

This reply goes into more detail.
https://github.com/AdamRamberg/unity-atoms/issues/23#issuecomment-482926060

**Comment by soraphis** [comment]
I opened the unity atoms project. Created a file at `Assets/Gizmos/UnityAtoms/FloatVariable Icon.png`

and as you can see in this image:

![image](https://user-images.githubusercontent.com/3066582/59564011-a2cb0980-9041-11e9-9bb5-92bcda6d17ea.png)

it obviosly works for assets created from scriptable objects. its not in the scene view, it does not have to be manually assigned

**Comment by AdamRamberg** [comment]
Thanks @jeffcampbellmakesgames and @soraphis for the input! 

Implemented a solution in cb6280a2554c9ac536d17485865fc4ae4ea85138 and bcdb8f16ea356fde0966260c9c4d60d7f56e2eb0.

I actually ended up doing something a bit more complicated, but more robust when it comes to assigning icons to generated atoms (using the Generator.cs). I took some inspiration from [this](https://stackoverflow.com/questions/40933372/programmatically-setting-and-saving-the-icon-associated-with-an-imported-asset) post on StackOverflow and did some adjustments to make it more flexible and general. It uses AssetPostprocessor to assign icons after assets are imported. Will probably break this out into its own repo, since I think that others could find it pretty nifty. 

Think we are good for know if comes to the icons. Might come back to this later, but closing the issue for now.
